### PR TITLE
Priv 7840 handling fingerprint will be lost cases

### PR DESCRIPTION
**Pull request by yoklmnprst**

### CODE REVIEW

File: Sources/PrivoSDK/PrivoAgeGate.swift [hunk 1]
```diff
@@ -15,8 +15,16 @@ public class PrivoAgeGate {
...
+    init(permissionService: PrivoPermissionServiceType = PrivoPermissionService.shared,
+         api: IRest = Rest.shared,
+         app: UIApplication = .shared,
+         fpidService: IFpidService = FpidService()
+    ) {
+        ageGate = PrivoAgeGateInternal(
+            permissionService: permissionService,
+            api: api,
+            app: app,
+            fpidService: fpidService)
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/PrivoAgeGate.swift:27]: wrong naming. Should be fpIdService

File: Sources/PrivoSDK/PrivoAuth.swift [hunk 2]
```diff
@@ -96,14 +96,20 @@ public class PrivoAuth {
     
     private let app: UIApplication
     private let userDefaults: UserDefaults
-    private let api: Rest
+    private let api: IRest
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/PrivoAuth.swift:99]: we do not use pattern with I prefixes in protocols, etc

File: Sources/PrivoSDK/api/Rest.swift [hunk 3]
```diff
@@ -40,12 +40,29 @@ extension URL {
...
+    func getAgeVerification(verificationIdentifier: String) async -> AgeVerificationTO?
+    func processLinkUser(data: LinkUserStatusRecord) async -> AgeGateStatusResponse?
+    func processBirthDate(data: FpStatusRecord) async throws -> AgeGateActionResponse?
+    func processRecheck(data: RecheckStatusRecord) async throws -> AgeGateActionResponse?
+    func trackCustomError(_ errorDescr: String)
+    func sendAnalyticEvent(_ event: AnalyticEvent)
+}
 
-class Rest {
+class Rest: IRest {
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/api/Rest.swift:61]: we really dislike I pattern (IRest, etc)

File: Sources/PrivoSDK/internal/PrivoInternal.swift [hunk 4]
```diff
@@ -10,7 +10,7 @@ internal class PrivoInternal {
     
     //MARK: - Static properties
         
-    static let rest = Rest()
+    static let rest: IRest = Rest.shared
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/PrivoInternal.swift:13]: do not like singletons patter. It's really outdated

Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/PrivoInternal.swift:13]: I absolutely agree with you, but this line only adds a protocol to an already existing solution. I think we will make an effort to get away from all singletons in the future.

File: Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift [hunk 5]
```diff
@@ -76,14 +74,59 @@ internal class AgeGateStorage {
...
+    var fpid: String? { get set }
+}
+
+protocol IFpidService {
+    var fpid: String? { get async }
+}
+
+class FpidService: IFpidService {
+    private let source: IRest
+    private var cache: IFpidStorage
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift:103]: let's call it storage, not cache

File: Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift [hunk 6]
```diff
@@ -76,14 +74,59 @@ internal class AgeGateStorage {
...
+    private func getFpId() async -> String? {
+        if let fpid = cache.fpid {
+            return fpid
+        }
+        
+        let rawFpid = DeviceFingerprint()
+        let fpidResponse = await source.generateFingerprint(fingerprint: rawFpid)
+        
+        guard let fpid = fpidResponse?.id else {
+            return nil
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift:126]: looks like for situations when we don't have FP ID we need to send AnalitycEvent

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeGateInternal.swift [hunk 7]
```diff
@@ -88,39 +103,34 @@ internal class PrivoAgeGateInternal {
...
-        do {
-            let settings = try await storage.serviceSettings.getSettings()
-            guard let settings = settings else { return nil }
-            let state = AgeState(fpId: fpId, agId: agId, settings: settings)
-            return state
-        } catch _ {
+        guard let settings = try? await storage.serviceSettings.getSettings() else {
             return nil
         }
+        let state = AgeState(agId: agId, settings: settings)
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/PrivoAgeGateInternal.swift:126]: From where you are getting settings now?

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift [hunk 8]
```diff
@@ -6,6 +6,7 @@
 //
 
 import Foundation
+import Alamofire
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift:9]: why are you moving Alamofire and network-check outside of the Network layer?

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift [hunk 9]
```diff
@@ -111,7 +112,10 @@ internal class PrivoAgeHelpers {
     }
     
     func checkNetwork() throws {
-        try PrivoInternal.rest.checkNetwork()
+        let rManager = NetworkReachabilityManager()
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift:115]: why is outside network layer ?

File: Sources/PrivoSDK/model/AgeGateData.swift [hunk 10]
```diff
@@ -34,7 +33,6 @@ internal struct CheckAgeStoreData: Encodable {
          birthDateYYYY: String?,
          redirectUrl: String?,
          agId: String?,
-         fpId: String?,
```
[the remark this answers is not in the document]
Comment (yoklmnprst) [reply Sources/PrivoSDK/model/AgeGateData.swift:37]: Really it is not used.

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeGateInternal.swift (hunk 7 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/PrivoAgeGateInternal.swift:126]: This fragment of code doesn't change any logic with settings: ```storage.serviceSettings.getSettings()``` used for getting value as it was.

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift (hunk 8 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift:9]: Okay. Removed the alamofire import here.

The main reason to move out checkNetwork from the Rest class due to these points:
- method checkNetwork is not tied to Rest instance data really
-  privo age helper instance have a dependency to rest fictitious instance 

Anyway, improved decision and delete this flashing dependency import.

File: Sources/PrivoSDK/PrivoAgeGate.swift (hunk 1 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/PrivoAgeGate.swift]: changed

File: Sources/PrivoSDK/PrivoAuth.swift (hunk 2 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/PrivoAuth.swift]: renamed

File: Sources/PrivoSDK/api/Rest.swift (hunk 3 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/api/Rest.swift]: renamed

File: Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift (hunk 5 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift]: okey, done

File: Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift (hunk 6 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift:126]: Thought about it and was ready to implement it, but stopped. And this is the reason:
```
    private func getFpId() async -> String? {
        if let fpId = storage.fpId {
            return fpId
        }
        
        let rawFpId = DeviceFingerprint()
        let fpIdResponse = await source.generateFingerprint(fingerprint: rawFpId)
        
        guard let fpId = fpIdResponse?.id else {
            // should sent analytic error here, but...
            return nil
        }
        storage.fpId = fpId

        return fpId
    }
```
We should be sent an analytic error if fpId returns nil. But
```
    func generateFingerprint(fingerprint: DeviceFingerprint) async -> DeviceFingerprintResponse? {
        let url = PrivoInternal.configuration.authBaseUrl.appending(.api).appending(.v1_0).appending(.fingerprint)
        typealias R = DataResponse<DeviceFingerprintResponse,AFError>
        let result: R = await session.request(url, method: .post, parameters: fingerprint, encoder: JSONParameterEncoder.default)
        // ... but if something really happened, try trackPossibleAFError already send the analytic error
        trackPossibleAFError(result.error,  result.debugDescription, result.response?.statusCode)
        return result.value
    }
```

if we get it wrong, trackPossibleAFError already did this work.

File: Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift (hunk 9 above)
Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/PrivoAgeHelpers.swift]: noticed above (previous thread)

File: Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift [hunk 11]
```diff
@@ -76,14 +74,59 @@ internal class AgeGateStorage {
...
+
+protocol FpIdStorage {
+    var fpId: String? { get set }
+}
+
+protocol FpIdable {
+    var fpId: String? { get async }
+}
+
+class FpIdService: FpIdable {
```
Comment (AlexSlo-Dev) [inline Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift:101]: looks like we need to move it to a separate file. It's not a storage anymore

Comment (yoklmnprst) [reply Sources/PrivoSDK/internal/age-gate/AgeGateStorage.swift]: moved it

File: Sources/PrivoSDK/model/AgeGateData.swift (hunk 10 above)
[the remark this answers is not in the document]
Comment (AlexSlo-Dev) [reply Sources/PrivoSDK/model/AgeGateData.swift:37]: AgeGateData is a model that we sent to Webview, right? Why do you decide that we don't use fp-id inside webview?

**Review by AlexSlo-Dev — APPROVED** [review]

**Pull request merged**